File: e17/proto/etk_server/examples/etk.pl
use IO::Socket;

system("etk_server &");
sleep(1);

my $sock = new IO::Socket::INET (
    PeerAddr => '127.0.0.1',
    PeerPort => '8080',
    Proto => 'tcp',
    Reuse => 1,
);
die "Could not create socket: $!\n" unless $sock;

sub etk_async
{
    print "a $_[0] $_[1]\0\n";
    print $sock "a $_[0] $_[1]\0";
    return $_[0];
}

sub etk_sync
{
    my $line;
    my $listener;
    
    print "s $_[0]\0\n";
    print $sock "s $_[0]\0";
    print "reading...\n";
    $listener = $sock->accept();
    $line = <$listener>;
    return $line;
}

my $app_id = etk_sync("0 etk_server_init");
  
my $win = etk_async("win1", "etk_window_new");
  
my $button1 = etk_async("btn1", "etk_button_new_with_label \"Perl Rules\"");

my $vbox = etk_async("vbx1", "etk_vbox_new 0 0");
etk_async("0", "etk_box_pack_start $vbox $button1 1 1 0");

etk_async("0", "etk_container_add $win $vbox");
  
etk_async("0", "etk_widget_show_all $win");

etk_async("0", "etk_server_signal_connect \"clicked\" $button1 \"button_1_click\"");
etk_async("0", "etk_server_signal_connect \"delete_event\" $win \"win_delete\"");
  
$event = "";
while ( $event ne "win_delete" )
{
    $event = etk_sync("0 etk_server_callback");
    
    if($event =~ /\s/)
    {
	($event, $tmp) = split " ", $event, 2;
	if($tmp =~ /(num|str)="(.*?)(?<!\\)"/)
	{
	    # todo, fill in args for event handler
	}
    }
    
    if($event eq "button_1_click")
    {
	print "button 1 clicked!\n";
    }
}

etk_async("0", "etk_main_quit");
etk_async("0", "etk_server_shutdown");

File: e17/proto/etk_server/examples/etk.java
import java.io.*;

public class etk {

  private static String FIFO = "/tmp/etk_server_fifo";

  private static String sendCommand(String command) throws IOException {
    /* print null terminated string */
    FileOutputStream out = new FileOutputStream(FIFO);
    out.write(command.getBytes());
    out.write('\0');
    out.close();

    /* read null terminated string (-1's may follow the termination too */
    FileInputStream in = new FileInputStream(FIFO);
    int inChar;
    StringBuffer ret = new StringBuffer();
    while ((inChar = in.read()) > 0)
	ret.append((char) inChar);
    in.close();
    return ret.toString();
  }

  public static void main(String args[]) throws IOException {
     
    System.out.println("Deprecated for now!");
    return;
    
    Runtime.getRuntime().exec("etk_server " + FIFO);
    System.out.println("running");
    try {
      Thread.sleep(1000);
    } catch (InterruptedException e) {
    }

    sendCommand("etk_init");
    String window = sendCommand("etk_window_new");
    String button = sendCommand("etk_button_new_with_label \"Java Rules\"");
    String vbox = sendCommand("etk_vbox_new 0 0");
    sendCommand("etk_box_pack_start " + vbox + " " + button + " 1 1 0");
    sendCommand("etk_container_add " + window + " " + vbox);
    sendCommand("etk_widget_show_all " + window);

    sendCommand("etk_server_signal_connect \"clicked\" " + button + " \"button_1_click\"");
    sendCommand("etk_server_signal_connect \"delete_event\" " + window + " \"win_delete\"");

    String event = "";
    while (!event.equals("win_delete")) {
      event = sendCommand("etk_server_callback");

      if (event.equals("button_1_click"))
        System.out.println("Button 1 clicked!");
    }

    sendCommand("etk_main_quit");
    sendCommand("etk_server_shutdown");
  }

}
